sycl/doc/DeviceGlobal.md

```diff
@@ -0,0 +1,658 @@
+
+```
+namespace sycl::detail {
+namespace {
+
+__sycl_device_global_registration::__sycl_device_global_registration() noexcept {
+  device_global_map::add(&::Foo,
+    /* mangled name of '::Foo' with unique suffix appended */,
+    /* size of underlying 'T' type */,
+    /* bool telling whether variable has 'device_image_scope` property */);
+  device_global_map::add(&::inner::Bar,
+    /* mangled name of '::inner::Bar' */,
+    /* size of underlying 'T' type */,
+    /* bool telling whether variable has 'device_image_scope` property */);
+}
+
+} // namepsace (unnamed)
+} // namespace sycl::detail
+```
+
+Note that a SYCL application can legally call SYCL APIs even before `main()` by
+calling them from a global constructor.  However, the integration headers have
+been designed to ensure that the address of each device global variable is
+registered with the DPC++ runtime before the user's application could legally
+use the variable, even if that use occurs before `main()` executes.
+
+The user's application cannot legally use a device global variable until the
+variable's constructor has been called, otherwise the application would be
+using an unconstructed object which has undefined behavior by C++ rules.  Since
+all device globals must be defined at namespace scope, the C++ rules for the
+order of global constructors only guarantee that the device global will be
+constructed before subsequent global variables in the same translation unit.
+Therefore, a user application could reference a device global from another
+global constructor only if that global constructor is for an object defined
+*after* the device global in the same translation unit.  However, the
+integration header defines `__sycl_device_global_registerer` *before* all
+device globals in the user's translation unit.  Therefore, the address of all
+device global variables in the translation unit will be registered with the
+DPC++ runtime before any user code could legally use them.
+
+#### Handling shadowed device global variables
+
+The example above shows a simple case where the user's device global variables
+can all be uniquely referenced via fully qualified lookup (e.g.
+`::inner::Bar`).  However, it is possible for users to construct applications
+where this is not the case, for example:
+
+```
+sycl::device_global<int> FuBar;
+namespace {
+  sycl::device_global<int> FuBar;
+}
+```
+
+In this example, the `FuBar` variable in the global namespace shadows a
+variable with the same name in the unnamed namespace.  The integration footer
+could reference the variable in the global namespace as `::FuBar`, but there is
+no way to reference the variable in the unnamed namespace using fully qualified
+lookup.
+
+Such programs are still legal, though.  The integration footer can support
+cases like this by defining a temporary variable that holds the address of the
+shadowed device global:
+
+```
+namespace {
+const void *__sycl_UNIQUE_STRING = &FuBar;  // References 'FuBar' in the
```

Comment:
@elizabethandrews, those two implementation are similar, but still different because of how we propagate information from the compiler to the runtime. For spec constants, we use device image properties and they are computed later in the toolchain, simply because we have no way of doing so earlier and therefore can't use integration footer like it is used for this feature.

So, here we emit code, which fills a map from a variable address to all metadata we need to attach to it.
For spec constants, we emit a string, which is used by the runtime to access the right device image property to extract the same metadata.

We could have used similar approach with registration functions and map for spec constants as well: to map variable address to a string identifier, but that would mean runtime dispatch (from address to a string), whilst the current implementation provides compile-time dispatch because it uses template specialization mechanism.

Reply:
Thank you for the information!